File: services/xstockstrat-analysis/app/telemetry.py
"""OpenTelemetry initialisation for xstockstrat-analysis.

Activated only when OTEL_ENABLED=true. All imports are deferred so the
service starts cleanly if the OTel packages are absent.
"""
import logging
import os

log = logging.getLogger(__name__)


def init_telemetry() -> None:
    """Configure OTel TracerProvider and instrument gRPC. No-op when disabled."""
    if os.getenv("OTEL_ENABLED", "false").lower() != "true":
        return

    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
        from opentelemetry.instrumentation.grpc import GrpcAioInstrumentorServer
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor

        endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "otel-collector:4317")
        svc_name = os.getenv("OTEL_SERVICE_NAME", "xstockstrat-analysis")
        environment = os.getenv("APP_ENV", "dev")

        resource = Resource.create(
            {
                "service.name": svc_name,
                "deployment.environment": environment,
            }
        )
        exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
        provider = TracerProvider(resource=resource)
        provider.add_span_processor(BatchSpanProcessor(exporter))
        trace.set_tracer_provider(provider)

        GrpcAioInstrumentorServer().instrument()

        log.info("OTel tracing initialised", extra={"endpoint": endpoint, "service": svc_name})

    except Exception as exc:  # noqa: BLE001
        log.warning("OTel init failed — continuing without tracing: %s", exc)

File: services/xstockstrat-insights/src/telemetry.ts
/**
 * OpenTelemetry initialisation for xstockstrat-insights.
 * Activated only when OTEL_ENABLED=true.
 * Must be called before any other imports to ensure auto-instrumentation works.
 */
export function initTelemetry(): void {
  if (process.env.OTEL_ENABLED !== 'true') return;

  try {
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { NodeSDK } = require('@opentelemetry/sdk-node');
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { OTLPTraceExporter } = require('@opentelemetry/exporter-trace-otlp-http');
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { resourceFromAttributes } = require('@opentelemetry/resources');
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { ATTR_SERVICE_NAME } = require('@opentelemetry/semantic-conventions');

    const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT ?? 'http://otel-collector:4318';
    const serviceName = process.env.OTEL_SERVICE_NAME ?? 'xstockstrat-insights';

    const sdk = new NodeSDK({
      resource: resourceFromAttributes({
        [ATTR_SERVICE_NAME]: serviceName,
        'deployment.environment': process.env.APPLICATION_ENV ?? 'development',
        trading_mode: process.env.TRADING_MODE ?? 'paper',
        platform: 'xstockstrat',
      }),
      traceExporter: new OTLPTraceExporter({ url: endpoint }),
      instrumentations: [],
    });

    sdk.start();
    console.info(`[otel] tracing enabled → ${endpoint} (service=${serviceName})`);

    process.on('SIGTERM', () => sdk.shutdown().catch(console.error));
  } catch (err) {
    console.warn('[otel] init failed — continuing without tracing:', err);
  }
}
